processor/src/main/java/org/mapstruct/ap/internal/model/source/Mapping.java

```diff
@@ -61,25 +56,51 @@
     private final NullValueCheckStrategyPrism nullValueCheckStrategy;
     private final NullValuePropertyMappingStrategyPrism nullValuePropertyMappingStrategy;
 
+    private final InheritContext inheritContext;
     private SourceReference sourceReference;
     private TargetReference targetReference;
 
-    public static Map<String, List<Mapping>> fromMappingsPrism(MappingsPrism mappingsAnnotation,
-        ExecutableElement method, FormattingMessager messager, Types typeUtils) {
-        Map<String, List<Mapping>> mappings = new HashMap<>();
+    public static class InheritContext {
+
+        private final boolean isReversed;
+        private final boolean isForwarded;
+        private final Method inheritedFromMethod;
+
+        public InheritContext(boolean isReversed, boolean isForwarded, Method inheritedFromMethod) {
+            this.isReversed = isReversed;
+            this.isForwarded = isForwarded;
+            this.inheritedFromMethod = inheritedFromMethod;
+        }
+
+        public boolean isReversed() {
+            return isReversed;
+        }
+
+        public boolean isForwarded() {
+            return isForwarded;
+        }
+
+        public Method getInheritedFromMethod() {
+            return inheritedFromMethod;
+        }
+    }
+
+    public static Set<String> getMappingTargetNamesBy(Predicate<Mapping> predicate, Set<Mapping> mappings) {
+        return mappings.stream().filter( predicate ).map( Mapping::getTargetName ).collect( Collectors.toSet() );
+    }
+
+    public static Mapping getMappingByTargetName(String targetName, Set<Mapping> mappings) {
+        return mappings.stream().filter( mapping -> mapping.targetName.equals( targetName ) ).findAny().orElse( null );
+    }
+
+    public static Set<Mapping> fromMappingsPrism(MappingsPrism mappingsAnnotation, ExecutableElement method,
+                                                 FormattingMessager messager, Types typeUtils) {
+        Set<Mapping> mappings = new LinkedHashSet<>();
 
         for ( MappingPrism mappingPrism : mappingsAnnotation.value() ) {
             Mapping mapping = fromMappingPrism( mappingPrism, method, messager, typeUtils );
             if ( mapping != null ) {
-                List<Mapping> mappingsOfProperty = mappings.get( mappingPrism.target() );
-                if ( mappingsOfProperty == null ) {
-                    mappingsOfProperty = new ArrayList<>();
-                    mappings.put( mappingPrism.target(), mappingsOfProperty );
-                }
-
-                mappingsOfProperty.add( mapping );
-
-                if ( mappingsOfProperty.size() > 1 && !isEnumType( method.getReturnType() ) ) {
+                if ( !mappings.add( mapping ) ) {
```

Comment:
Maybe for making it more clear we should do

```
if ( mappings.contains(mapping) ) {
...
} else {
    mappings.add( mapping );
}
```

Otherwise you always need to read what the return of the `add` means. I think that it is easier to read it like this